daemon/cmd/kubeproxy_replacement_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux,privileged_tests
+
+package cmd
+
+import (
+	"net"
+	"runtime"
+	"sort"
+
+	"github.com/cilium/cilium/pkg/checker"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	. "gopkg.in/check.v1"
+)
+
+type KubeProxySuite struct {
+	currentNetNS                  netns.NsHandle
+	testNetNS                     netns.NsHandle
+	prevConfigDevices             []string
+	prevConfigDirectRoutingDevice string
+	prevConfigEnableIPv4          bool
+	prevConfigEnableIPv6          bool
+	prevK8sNodeIP                 net.IP
+}
+
+var _ = Suite(&KubeProxySuite{})
+
+func (s *KubeProxySuite) SetUpSuite(c *C) {
+	var err error
+
+	s.prevConfigDevices = option.Config.Devices
+	s.prevConfigDirectRoutingDevice = option.Config.DirectRoutingDevice
+	s.prevConfigEnableIPv4 = option.Config.EnableIPv4
+	s.prevConfigEnableIPv6 = option.Config.EnableIPv6
+	s.prevK8sNodeIP = node.GetK8sNodeIP()
+
+	s.currentNetNS, err = netns.Get()
```

Comment:
These operations likely need to be protected by locking the goroutine to the thread via `runtime.LockOSThread`. See [netns's docs](https://pkg.go.dev/github.com/vishvananda/netns?tab=doc) for more information. See [this commit](9ded3a3d743a1771291e8561f215c6b7b3725d6d) for an example.

Relevant snippet:

> Package netns allows ultra-simple network namespace handling. NsHandles can be retrieved and set. Note that the current namespace is thread local so actions that set and reset namespaces should use LockOSThread to make sure the namespace doesn't change due to a goroutine switch. It is best to close NsHandles when you are done with them. This can be accomplished via a `defer ns.Close()` on the handle. Changing namespaces requires elevated privileges, so in most cases this code needs to be run as root.

Reply:
Moving the locking to `SetUpTest` is IMHO error prone, as it's up to a testing framework to decide whether to run each test case in the same goroutine, and this might be not visible to a developer. If it end up running the test case in a different goroutine, then subtle bugs might be introduced, see https://www.weave.works/blog/linux-namespaces-golang-followup on this topic.